### PR TITLE
Added  cartesian gps coords relative home point as additional computed fields

### PR DESCRIPTION
**Pull request by demvlad**
The log files can include gps coords - latitude, longitude, altitude.
But direct using of raw gps data is not comfortable.
Therefore this PR add computed fields - cartesian coords around home point to show its at the chart:
"gpsCartesianCoords[0]" -  X coord (axis x direction is North)
"gpsCartesianCoords[1]",   Y coord - gps altitude
"gpsCartesianCoords[2]"-  Z coord (axis z direction is East)
"gpsDistance" - distance from home point

The Y coord is absolute GPS altitude at moment. Because we need gps home altitude to compute relative altitude. But it is not exists in log at current time.
When gps home altitude will added to log in BF firmware, the Y coord will show relative altitude around home point.
The BF PR what add logging of gps home altitude is ready and can be published if this PR will aproved.

### CODE REVIEW

File: src/flightlog_fields_presenter.js [hunk 1]
```diff
@@ -124,6 +124,12 @@ const FRIENDLY_FIELD_NAMES = {
   GPS_altitude: "GPS Altitude ASL",
   GPS_speed: "GPS Speed",
   GPS_ground_course: "GPS Heading",
+  
+  "gpsCartesianCoords[all]": "GPS Coords",
+  "gpsCartesianCoords[0]": "GPS Coords [X]",
+  "gpsCartesianCoords[1]": "GPS Coords [H]",
```
Comment (haslinghuis) [inline src/flightlog_fields_presenter.js:130]:
```suggestion
  "gpsCartesianCoords[1]": "GPS Coords [Y]",
```

File: src/flightlog.js [hunk 2]
```diff
@@ -519,9 +524,9 @@ export function FlightLog(logData) {
                 }
                 break;
               case "H":
-              // TODO
-              // contains coordinates only
-              // should be handled separately
+                const homeAltitude = frame.length > 2 ? frame[2]/10 : 0; // will work after BF firmware improvement
+                gpsTransform = new GPS_transform(frame[0]/10000000, frame[1]/10000000, homeAltitude, 0.0);
```
Comment (haslinghuis) [inline src/flightlog.js:528]:
```suggestion
                const homeAltitude = frame.length > 2 ? frame[2 ] /10 : 0; // will work after BF firmware improvement
                gpsTransform = new GPS_transform(frame[0] / 10000000, frame[1] / 10000000, homeAltitude, 0.0);
```

File: src/flightlog.js [hunk 3]
```diff
@@ -830,6 +845,22 @@ export function FlightLog(logData) {
             }
           }
 
+          // Calculate cartesian coords by GPS
+          if (!that.isFieldDisabled().GPS) {
+            if (gpsTransform && gpsCoord && srcFrame[gpsCoord[0]]) {
+              const gpsCartesianCoords = gpsTransform.WGS_BS(srcFrame[gpsCoord[0]]/10000000, srcFrame[gpsCoord[1]]/10000000, srcFrame[gpsCoord[2]]/10);
```
Comment (haslinghuis) [inline src/flightlog.js:851]:
```suggestion
              const gpsCartesianCoords = gpsTransform.WGS_BS(srcFrame[gpsCoord[0]] / 10000000, srcFrame[gpsCoord[1] ] /10000000, srcFrame[gpsCoord[2]] / 10);
```

File: src/flightlog.js [hunk 4]
```diff
@@ -830,6 +845,22 @@ export function FlightLog(logData) {
...
+              destFrame[fieldIndex++] = gpsCartesianCoords.x;
+              destFrame[fieldIndex++] = gpsCartesianCoords.y;
+              destFrame[fieldIndex++] = gpsCartesianCoords.z;
+              destFrame[fieldIndex++] = Math.sqrt(gpsCartesianCoords.x*gpsCartesianCoords.x + gpsCartesianCoords.z*gpsCartesianCoords.z);
```
Comment (haslinghuis) [inline src/flightlog.js:855]:
```suggestion
              destFrame[fieldIndex++] = Math.sqrt(gpsCartesianCoords.x * gpsCartesianCoords.x + gpsCartesianCoords.z * gpsCartesianCoords.z);
```

File: src/gps_transform.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      x: c11*PosX1 + c12*PosZ1,
+      y: a31*(pos.x-X0) + a32*(pos.y-Y0) + a33*(pos.z-Z0),
+      z: c21*PosX1 + c22*PosZ1,
+    };
+  };
+
+  this.WGS_BS = function (Lat, Lon, H) {
+    return this.ECEF_BS(this.WGS_ECEF(Lat, Lon, H));
+  };
+}
```
Comment (haslinghuis) [inline src/gps_transform.js:67]: Please add spaces around operators for better readability
```suggestion

}
```

**Comment by demvlad** [comment]
GPS Home altitude is added in this [BF PR](https://github.com/betaflight/betaflight/pull/13939)

File: src/gps_transform.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      y: a31*(pos.x-X0) + a32*(pos.y-Y0) + a33*(pos.z-Z0),
+      z: c21*PosX1 + c22*PosZ1,
+    };
+  };
+
+  this.WGS_BS = function (Lat, Lon, H) {
+    return this.ECEF_BS(this.WGS_ECEF(Lat, Lon, H));
+  };
+
+}
```
Comment (haslinghuis) [inline src/gps_transform.js:68]: Sorry was meant to be 

```suggestion
}

Also please add spaces around operators in this file.
```

File: src/flightlog.js [hunk 7]
```diff
@@ -519,9 +524,9 @@ export function FlightLog(logData) {
                 }
                 break;
               case "H":
-              // TODO
-              // contains coordinates only
-              // should be handled separately
+                const homeAltitude = frame.length > 2 ? frame[2 ] /10 : 0; // will work after BF firmware improvement
```
Comment (haslinghuis) [inline src/flightlog.js:527]: need to declare with let before the switch case and assign value here.

Comment (demvlad) [reply src/flightlog.js]: What result will ins this variable with different frames type? We need it for "H" type only.
The SonarCloud shows the issue, but in my opinion, this is more good place to define it.
I've tryed to put this expression in the next code rows directly, without variable. But SonarCloud shows the issue too.

**Review by ctzsnooze — DISMISSED** [review]

File: src/flightlog.js (hunk 7 above)
Comment (haslinghuis) [reply src/flightlog.js]: Sonar won't complain if you wrap the case in brackets.

Comment (demvlad) [reply src/flightlog.js]: Yes, it works.

**Review by haslinghuis — APPROVED** [review]

**Pull request merged**